io/slow5/slow5.go

```diff
@@ -0,0 +1,409 @@
+		}
+	}
+	// Now that we know what attribute values are possible, lets build a map
+	// with those values with "." placeholders (as defined in slow5 spec)
+	headerAttributes := make(map[string][]string)
+	for key := range possibleAttributeKeys {
+		newBlankAttributes := make([]string, len(headers))
+		for blankAttributeIndex := range newBlankAttributes {
+			newBlankAttributes[blankAttributeIndex] = "."
+		}
+		headerAttributes[key] = newBlankAttributes
+	}
+	// Build a list with all header values
+	var headerAttributeStrings []string
+	for headerIndex, header := range headers {
+		for key, value := range header.Attributes {
+			headerAttributes[key][headerIndex] = value
+		}
+	}
+	for key, value := range headerAttributes {
+		var attributeRow strings.Builder
+		attributeRow.Write([]byte(key))
+		for _, attributeValue := range value {
+			attributeRow.Write([]byte("\t"))
+			attributeRow.Write([]byte(attributeValue))
+		}
+		headerAttributeStrings = append(headerAttributeStrings, attributeRow.String())
+	}
+	// Sort the header strings
+	sort.Strings(headerAttributeStrings)
+
+	// Write the header attribute strings to the output
+	for _, headerAttributeString := range headerAttributeStrings {
+		_, err = output.Write([]byte(fmt.Sprintf("%s\n", headerAttributeString)))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Write the read headers
+	// These are according to the slow5 specifications
+	_, err = output.Write([]byte("#char*	uint32_t	double	double	double	double	uint64_t	int16_t*	uint64_t	int32_t	uint8_t	double	enum{unknown,partial,mux_change,unblock_mux_change,data_service_unblock_mux_change,signal_positive,signal_negative}	char*\n"))
+	if err != nil {
+		return err
+	}
+	_, err = output.Write([]byte("#read_id	read_group	digitisation	offset	range	sampling_rate	len_raw_signal	raw_signal	start_time	read_number	start_mux	median_before	end_reason	channel_number\n"))
+	if err != nil {
+		return err
+	}
+	// Parse failure reasons
+	failureReasonsString := "unknown,partial,mux_change,unblock_mux_change,data_service_unblock_mux_change,signal_positive,signal_negative"
+	failureReasons := make(map[string]int)
+	for failureIndex, failureReason := range strings.Split(failureReasonsString, ",") {
+		failureReasons[failureReason] = failureIndex
+	}
+
+	// Iterate over reads. This is reading from a channel, and will end
+	// when the channel is closed.
+	for read := range reads {
+		// converts []int16 to string
+		rawSignalString := strings.Trim(strings.Replace(fmt.Sprint(read.RawSignal), " ", ",", -1), "[]")
+		// Look at above output.Write("#read_id ... for the values here.
+		_, err = output.Write([]byte(fmt.Sprintf("%s\t%d\t%g\t%g\t%g\t%g\t%d\t%s\t%d\t%d\t%d\t%g\t%d\t%s\n", read.ReadId, read.ReadGroupId, read.Digitisation, read.Offset, read.Range, read.SamplingRate, read.LenRawSignal, rawSignalString, read.StartTime, read.ReadNumber, read.StartMux, read.MedianBefore, failureReasons[read.EndReason], read.ChannelNumber)))
```

Comment:
Suggest using fmt.Fprintf instead of the `Write` method- might save you quite a bit of allocations depending on how well the compiler optimized the current line i.e.:
```go
_, err = fmt.Fprintf(output, "%s\t%d\t%g\t%g\t%g\t%g\t%d\t%s\t%d\t%d\t%d\t%g\t%d\t%s\n", read.ReadId, //...
```